### PR TITLE
object avoidance project + I2C mutex + updated docs

### PR DESCRIPTION
**Pull request by RobertLucian**
In this PR, I'm bringing a project built on top of the `GoPiGo3`. 

It's a small example project, where the robot is supposed to move forward as long as it's not too close to a target. Its speed is proportional to the distance from the target.

The following issues have also been fixed:

* https://github.com/RobertLucian/GoPiGo3/issues/10
* https://github.com/RobertLucian/GoPiGo3/issues/9
* https://github.com/RobertLucian/GoPiGo3/issues/7

----------

Along with this project, I've also modified the way the mutex on `I2C` works. Instead of using a function, there's now a class from which another class can derive. Also, the user can now set whether `DistanceSensor` and `DHTSensor` objects are multithreaded.

-----------

Here's a link to the [project](https://github.com/RobertLucian/GoPiGo3/tree/feature/object-avoidance-project/Projects/ObjectAvoidanceRobot). The link will be valid as long as the PR is open.

-----------

Before accepting this PR, it's also needed to bring the required modifications to the [GoPiGo2](https://github.com/DexterInd/GoPiGo) repository.

-----------

The documentation has been updated with the latest changes and it looks like this:

distance

### CODE REVIEW

**Comment by RobertLucian** [comment]
After a couple of small mods, I've tested both the `DistanceSensor` and `DHTSensor` classes and they worked as expected both when using the `mutex` and when not.

File: Software/Python/easygopigo3.py [hunk 1]
```diff
@@ -28,8 +28,6 @@
 except:
```
Comment (johnisanerd) [inline Software/Python/easygopigo3.py]: This is the second one I'm worried about.  I will defer to @CleoQc  on this one though.  We might also have @mattallen37  look at it.

File: Software/Python/I2C_mutex.py [hunk 2]
```diff
@@ -1,27 +1,56 @@
 import fcntl
```
Comment (johnisanerd) [inline Software/Python/I2C_mutex.py]: This is the First one I'm worried about.  I will defer to @CleoQc  on this one though.  We might also have @mattallen37  look at it.

Comment (RobertLucian) [inline Software/Python/I2C_mutex.py]: Why are you afraid of the `fcntl`? What are you thinking of?

File: Software/Python/easygopigo3.py [hunk 3]
```diff
@@ -2453,13 +2440,16 @@ def read_mm(self):
         # try 3 times to have a reading that is
         # smaller than 8m or bigger than 5 mm.
         # if sensor insists on that value, then pass it on
+
         while (mm > 8000 or mm < 5) and attempt < 3:
+            if self.mutex:
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: Put these and all others like it into one line

Comment (RobertLucian) [inline Software/Python/easygopigo3.py]: It's done and tested.

**Comment by RobertLucian** [comment]
I've solved the conflicts in this one. Apparently, the latest changes made this PR un-merge-able.

File: Software/Python/easygopigo3.py [hunk 4]
```diff
@@ -2756,13 +2798,9 @@ def read_humidity(self):
...
 
-        try:
-            _grab_read()
-            humidity = DHT.dht(self.sensor_type)[1]
-        except Exception as e:
-            raise
-        finally:
-            _release_read()
+        self.__ifMutexAcquire()
+        humidity = DHT.dht(self.sensor_type)[1]
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: This should be within Try/Except/Finally

File: Software/Python/easygopigo3.py [hunk 5]
```diff
@@ -2733,13 +2779,9 @@ def read_temperature(self):
...
 
-        try:
-            _grab_read()
-            temp = DHT.dht(self.sensor_type)[0]
-        except Exception as e:
-            raise
-        finally:
-            _release_read()
+        self.__ifMutexAcquire()
+        temp = DHT.dht(self.sensor_type)[0]
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: Try/Except/Finally

File: Software/Python/easygopigo3.py [hunk 6]
```diff
@@ -2644,14 +2650,12 @@ def read_mm(self):
...
+            self.__ifMutexAcquire()
             try:
-                _grab_read()
                 mm = self.read_range_single()
             except:
                 mm = 0
-            finally:
-                _release_read()
-                
+            self.__ifMutexRelease()
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: To be kind to our future self, the mutexRelease is under finally:

File: Software/Python/easygopigo3.py [hunk 7]
```diff
@@ -2724,6 +2728,48 @@ def __init__(self, port="SERIAL",gpg=None, sensor_type=0):
         except:
             raise
 
+        try:
+            import threading
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: is that new code?

File: Software/Python/easygopigo3.py [hunk 8]
```diff
@@ -2775,13 +2813,9 @@ def read_humidity(self):
...
         from di_sensors import DHT
 
-        try:
-            _grab_read()
-            [temp , humidity]=DHT.dht(self.sensor_type)
-        except Exception as e:
-            raise
-        finally:
-            _release_read()
+        self.__ifMutexAcquire()
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: try/except/finally

File: Software/Python/easygopigo3.py (hunk 7 above)
Comment (RobertLucian) [reply Software/Python/easygopigo3.py]: Really good catch. This an old remnant.

**Comment by RobertLucian** [comment]
@CleoQc so I've gone through the source code and I've noticed really peculiar things - nevertheless I've fixed them (probably because of the former conflicts things got messed up), added support for the DHT sensor in the documentation, polished a bit the DHT sensor and made some further changes to the documentation.

There's one last thing. The thing with the mutexes has been done the following way:
```
acquireMutex
try:
    do whatever
except:
    throw whatever
releaseMutex
```
How does this sound to you? My motive was that I didn't want to do the acquiring and releasing of mutexes inside the `try/except/finally` blocks. Even if the `finally` block gets executed no matter what, it feels and looks better to keep it out of it. What do you think? Do you have a specific reason to keep in there? I saw you said *to be kind to our future self* and I wonder what you're thinking about.

**Comment by RobertLucian** [comment]
@CleoQc I've tested it and I've added the requested changes so it's all good for a review now.

File: Software/Python/I2C_mutex.py [hunk 9]
```diff
@@ -1,27 +1,56 @@
...
-    acquired = False
-    while not acquired:
         try:
-            # lock
-            fcntl.lockf(DexterLockI2C_handle, fcntl.LOCK_EX | fcntl.LOCK_NB)
-            acquired = True
-        except IOError: # already locked by a different process
+            DexterLockI2C_handle = open('/run/lock/DexterLockI2C')
+        except:
+            pass
```
Comment (CleoQc) [inline Software/Python/I2C_mutex.py]: what may cause this to fail? Wouldn't it be best to at least print the exception?

File: Software/Python/easygopigo3.py [hunk 10]
```diff
@@ -2724,71 +2750,81 @@ def __init__(self, port="SERIAL",gpg=None, sensor_type=0):
...
         from di_sensors import DHT
 
-        try:
-            _grab_read()
-            humidity = DHT.dht(self.sensor_type)[1]
-        except Exception as e:
-            raise
-        finally:
-            _release_read()
+        _ifMutexAcquire(self.use_mutex)
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: this used to be in a try/except. Even if dht sensor doesn't actually raise an error, I think it's best to keep everything the same.

File: Software/Python/easygopigo3.py [hunk 11]
```diff
@@ -2724,71 +2750,81 @@ def __init__(self, port="SERIAL",gpg=None, sensor_type=0):
...
 
-        try:
-            _grab_read()
-            [temp , humidity]=DHT.dht(self.sensor_type)
-        except Exception as e:
-            raise
-        finally:
-            _release_read()
+        _ifMutexAcquire(self.use_mutex)
+        [temp , humidity]=DHT.dht(self.sensor_type)
```
Comment (CleoQc) [inline Software/Python/easygopigo3.py]: Same here. Please keep the try/except/finally structure

File: Software/Python/easygopigo3.py (hunk 10 above)
Comment (RobertLucian) [reply Software/Python/easygopigo3.py]: You mean about keeping the `ifMutexWhatever` inside the finally block? If so, I already did that, but it doesn't show up.

File: Software/Python/I2C_mutex.py (hunk 9 above)
Comment (RobertLucian) [reply Software/Python/I2C_mutex.py]: No. There are 2 reasons:

1. Doing it so will lead to a very ugly console output, provided someone does use multiple threads/processes. Basically, the console will be filled with the said message.
2. Second of all, this is considered to be a bad practice. If we would want to do something about it, we could create a timeout on this method (most probably through another specialized class) and then end the method when the time runs out. Another option is using modules that are intended to be used for logging the output. Bottom line is: we should steer away from prints.

Also, this fails whenever the resource is locked by another thread/process - I wrote there a comment.

Comment (CleoQc) [reply Software/Python/I2C_mutex.py]: Ok, you're right. All good then on this point

**Pull request merged**